File: similarity-service/internal/database/model/task_similarity.go
package model

import (
	"database/sql"
	"time"
)

type CloneLevel struct {
	CloneLevel     string          `json:"cloneLevel"`
	CloneLevelName string          `json:"cloneLevelName"`
	Similarity     sql.NullFloat64 `json:"similarity"`
	Count          int             `json:"count"`
	Percentage     float64         `json:"percentage"`
}

type TaskReportStats struct {
	SampleCount   int     `json:"sampleCount"`
	GroupCount    int     `json:"groupCount"`
	AvgSimilarity float64 `json:"avgSimilarity"`
	MaxSimilarity float64 `json:"maxSimilarity"`
}

// 检测结果任务库
type TaskSimilarity struct {
	ID               string     `gorm:"column:id;primaryKey;type:varchar(32)"`
	TaskID           string     `gorm:"column:task_id;type:varchar(32)"`
	TaskType         bool       `gorm:"column:task_type;default:false"`
	ProblemID        string     `gorm:"column:problem_id;type:varchar(32)"`
	SetID            string     `gorm:"column:set_id;type:varchar(32)"`
	IsSet            bool       `gorm:"column:is_set;default:false"`
	Language         string     `gorm:"column:language;type:varchar(64)"`
	Similarity       float64    `gorm:"column:similarity;type:decimal(10,2);default:0.00"`
	SubmitUser       string     `gorm:"column:submit_user;type:varchar(32)"`
	SubmitCode       string     `gorm:"column:submit_code;type:longtext"`
	SubmitCodeLength int        `gorm:"column:submit_code_length;default:0"`
	SubmitID         string     `gorm:"column:submit_id;type:varchar(32)"`
	SubmitTime       *time.Time `gorm:"column:submit_time"`
	SubmitCodeToken  string     `gorm:"column:submit_code_token;type:json"`
	SubmitTokenName  string     `gorm:"column:submit_token_name;type:json"`
	SubmitTokenTexts string     `gorm:"column:submit_token_texts;type:json"`
	OriginUser       string     `gorm:"column:origin_user;type:varchar(32)"`
	OriginCode       string     `gorm:"column:origin_code;type:longtext"`
	OriginCodeLength int        `gorm:"column:origin_code_length;default:0"`
	OriginID         string     `gorm:"column:origin_id;type:varchar(32)"`
	OriginTime       *time.Time `gorm:"column:origin_time"`
	OriginCodeToken  string     `gorm:"column:origin_code_token;type:json"`
	OriginTokenName  string     `gorm:"column:origin_token_name;type:json"`
	OriginTokenTexts string     `gorm:"column:origin_token_texts;type:json"`
	Deleted          bool       `gorm:"column:deleted;default:false"`
	CreateTime       *time.Time `gorm:"column:create_time"`
	CreateUser       string     `gorm:"column:create_user;type:varchar(32)"`
	UpdateTime       *time.Time `gorm:"column:update_time"`
	UpdateUser       string     `gorm:"column:update_user;type:varchar(32)"`
}

func (TaskSimilarity) TableName() string {
	return "task_similarity"
}

File: galaxy-web-app/galaxy-oj/src/main/java/io/charlie/web/oj/modular/data/similarity/entity/TaskSimilarity.java
package io.charlie.web.oj.modular.data.similarity.entity;

import com.baomidou.mybatisplus.annotation.TableField;
import com.baomidou.mybatisplus.annotation.TableName;
import com.baomidou.mybatisplus.annotation.TableName;
import com.baomidou.mybatisplus.extension.handlers.JacksonTypeHandler;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import io.charlie.galaxy.config.timestamp.DateToTimestampSerializer;
import io.charlie.galaxy.config.timestamp.TimestampToDateDeserializer;
import io.charlie.galaxy.pojo.CommonEntity;

import java.io.Serializable;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.io.Serial;
import java.util.Date;
import java.util.List;

import io.charlie.web.oj.modular.data.problem.entity.DataProblem;
import io.charlie.web.oj.modular.data.set.entity.DataSet;
import io.charlie.web.oj.modular.sys.user.entity.SysUser;
import io.swagger.v3.oas.annotations.media.Schema;
import lombok.Data;
import lombok.EqualsAndHashCode;
import org.dromara.core.trans.anno.Trans;
import org.dromara.core.trans.constant.TransType;

/**
 * @author Charlie Zhang
 * @version v1.0
 * @date 2025-09-21
 * @description 检测结果任务库
 */
@EqualsAndHashCode(callSuper = true)
@Data
@TableName(value = "task_similarity", autoResultMap = true)
@Schema(name = "TaskSimilarity", description = "检测结果任务库")
public class TaskSimilarity extends CommonEntity {
    @Serial
    @TableField(exist = false)
    private static final long serialVersionUID = 1L;

    @Schema(description = "主键")
    private String id;

    @Schema(description = "任务ID")
    private String taskId;

    @Schema(description = "手动")
    @Trans(type = TransType.DICTIONARY, key = "YES_NO")
    private Boolean taskType;

    @Schema(description = "题目ID")
    @Trans(type = TransType.SIMPLE, target = DataProblem.class, fields = "title", ref = "problemIdName")
    private String problemId;

    @Schema(description = "题目名称")
    @TableField(exist = false)
    private String problemIdName;

    @Schema(description = "题集ID")
    @Trans(type = TransType.SIMPLE, target = DataSet.class, fields = "title", ref = "setIdName")
    private String setId;

    @Schema(description = "题集名称")
    @TableField(exist = false)
    private String setIdName;

    @Schema(description = "是否是题集提交")
    @Trans(type = TransType.DICTIONARY, key = "YES_NO")
    private Boolean isSet;

    @Schema(description = "编程语言")
    @Trans(type = TransType.DICTIONARY, key = "ALLOW_LANGUAGE", ref = "languageName")
    private String language;

    @Schema(description = "语言名称")
    @TableField(exist = false)
    private String languageName;

    @Schema(description = "相似度")
    private BigDecimal similarity;

    @Schema(description = "提交用户")
    @Trans(type = TransType.SIMPLE, target = SysUser.class, fields = {"nickname", "avatar"}, refs = {"submitUserName", "submitUserAvatar"})
    private String submitUser;

    @Schema(description = "用户名称")
    @TableField(exist = false)
    private String submitUserName;

    @Schema(description = "用户头像")
    @TableField(exist = false)
    private String submitUserAvatar;

    @Schema(description = "源代码")
    private String submitCode;

    @Schema(description = "源代码长度")
    private Integer submitCodeLength;

    @Schema(description = "提交ID")
    private String submitId;

    @Schema(description = "提交时间")
    @JsonSerialize(using = DateToTimestampSerializer.class)
    @JsonDeserialize(using = TimestampToDateDeserializer.class)
    private Date submitTime;


    @TableField(typeHandler = JacksonTypeHandler.class)
    private List<Integer> submitCodeToken;

    @Schema(description = "提交用户Token名称")
    @TableField(typeHandler = JacksonTypeHandler.class)
    private List<String> submitTokenName;

    @Schema(description = "提交用户Token内容")
    @TableField(typeHandler = JacksonTypeHandler.class)
    private List<String> submitTokenTexts;

    @Schema(description = "样本用户")
    @Trans(type = TransType.SIMPLE, target = SysUser.class, fields = {"nickname", "avatar"}, refs = {"originUserName", "originUserAvatar"})
    private String originUser;

    @Schema(description = "用户名称")
    @TableField(exist = false)
    private String originUserName;

    @Schema(description = "用户头像")
    @TableField(exist = false)
    private String originUserAvatar;

    @Schema(description = "样本源代码")
    private String originCode;

    @Schema(description = "样本源代码长度")
    private Integer originCodeLength;

    @Schema(description = "样本提交ID")
    private String originId;

    @Schema(description = "样本提交时间")
    @JsonSerialize(using = DateToTimestampSerializer.class)
    @JsonDeserialize(using = TimestampToDateDeserializer.class)
    private Date originTime;

    @TableField(typeHandler = JacksonTypeHandler.class)
    private List<Integer> originCodeToken;

    @Schema(description = "样本用户Token名称")
    @TableField(typeHandler = JacksonTypeHandler.class)
    private List<String> originTokenName;

    @Schema(description = "样本用户Token内容")
    @TableField(typeHandler = JacksonTypeHandler.class)
    private List<String> originTokenTexts;
}
